server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -775,9 +780,13 @@ private static ActionListener<SearchResponse> createCCSListener(
         ) {
             @Override
             void innerOnResponse(SearchResponse searchResponse) {
-                // TODO: in CCS fail fast ticket we may need to fail the query if the cluster gets marked as FAILED
                 ccsClusterInfoUpdate(searchResponse, clusters, clusterAlias, skipUnavailable);
                 searchResponseMerger.add(searchResponse);
+                if (progressListener != SearchProgressListener.NOOP) {
+                    // do an incremental merge of all SearchResponses received so far for async-search with CCS MRT=true
+                    // (sync search uses SearchProgressListener.NOOP, so this path will not execute)
+                    progressListener.notifyCcsMinimizeRoundtripsReduce(searchResponseMerger.getMergedResponse(clusters));
```

Comment:
Ok I finally understood what you meant in another comment: we do hold an additional search response in memory, because search response merger still does what it's supposed to do and holds all the search responses from all remote clusters until they all come back. In addition, we do a partial merge to update the async search result.

An alternative if we want to save memory is to delay the merging until results are retrieved, similar to how we deal with partial aggs results currently. It does require some changes to your current PR. We could either to it every time, or cache the result after the first get async search call (like we do for aggs if I am not mistaken).